File: solutions/java/acronym/1/src/main/java/Acronym.java
import java.util.Arrays;
import java.util.function.Function;
import java.util.stream.Collectors;

public final class Acronym {
	private final String phrase;
	private String acronym;

	public Acronym(String phrase) {
		this.phrase = phrase;
	}

	public String get() {
		if (this.acronym == null) {
			this.acronym = createAcronym();
		}

		return this.acronym;
	}

	private String createAcronym() {
		return Arrays.stream(this.splitOnSpacesAndHyphens())
				       .map(this.removeNonLetters())
				       .filter(s -> !s.isEmpty())
				       .map(this.getFirstLetter())
				       .collect(Collectors.joining());
	}

	private String[] splitOnSpacesAndHyphens() {
		return this.phrase.split("[\\s-]+");
	}

	private Function<String, String> removeNonLetters() {
		return word -> word.replaceAll("[^\\p{L}]", "");
	}

	private Function<String, String> getFirstLetter() {
		return s -> String.valueOf(Character.toUpperCase(s.charAt(0)));
	}
}

File: solutions/ocaml/acronym/1/acronym.ml
open Base

let split_into_words phrase =
    phrase
    |> String.split_on_chars ~on:[' '; '-'; '_']

let is_not_empty_string = Fn.non String.is_empty

let get_first_letter word = String.get word 0

let get_first_letter_as_uppercase word =
    word
    |> get_first_letter
    |> Char.uppercase

let first_letters_uppercase words =
  words
  |> List.filter ~f:is_not_empty_string
  |> List.map ~f:get_first_letter_as_uppercase
  |> String.of_char_list

let acronym phrase =
  phrase
  |> split_into_words
  |> first_letters_uppercase
